cli/flags.go

```diff
@@ -0,0 +1,697 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//nolint:wrapcheck // These functions intentionally just wrap flag.Flag.
+package cli
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/abcxyz/pkg/timeutil"
+	"github.com/kr/text"
+)
+
+const maxLineLength = 80
+
+// LookupEnvFunc is the signature of a function for looking up environment
+// variables. It makes that of [os.LookupEnv].
+type LookupEnvFunc = func(string) (string, bool)
+
+// MapLookuper returns a LookupEnvFunc that reads from a map instead of the
+// environment. This is mostly used for testing.
+func MapLookuper(m map[string]string) LookupEnvFunc {
+	return func(s string) (string, bool) {
+		v, ok := m[s]
+		return v, ok
+	}
+}
+
+// FlagSet is the root flag set for creating and managing flag sections.
+type FlagSet struct {
+	flagSet   *flag.FlagSet
+	sections  []*FlagSection
+	lookupEnv LookupEnvFunc
+}
+
+// Option is an option to the flagset.
+type Option func(fs *FlagSet) *FlagSet
+
+// WithLookupEnv defines a custom function for looking up environment variables.
+// This is mostly useful for testing.
+func WithLookupEnv(fn LookupEnvFunc) Option {
+	if fn == nil {
+		panic("lookup cannot be nil")
+	}
+
+	return func(fs *FlagSet) *FlagSet {
+		fs.lookupEnv = fn
+		return fs
+	}
+}
+
+// NewFlagSet creates a new root flag set.
+func NewFlagSet(opts ...Option) *FlagSet {
```

Comment:
Maybe it's overkill. From the JVS change, I see the way to use flags with a command is common. So here is an idea:

```go
// Interface for commands with flags
type Flaggable interface {
  Flags() *FlagSet
}

// Have a wrapper for commands with flags
type LeafCommand struct {
  Command // Embed a command
}

// This handles concat the help string
func (c *LeafCommand) Help() string {
  var flagHelp string
  if fc, ok := c.Command.(Flaggable); ok {
    flagHelp = fc.Flags().Help()
  }
  return return strings.Trim(c.Command.Helper()+flagHelp)
}

// If the actual command implementation has flags,
// parse the flags and pass it over.
func (c *LeafCommand) Run(ctx context.Context, args []string) error {
  if fc, ok := c.Command.(Flaggable); ok {
    f := fc.Flags()
    if err := f.Parse(args); err != nil {
      return fmt.Errorf("failed to parse flags: %w", err)
    }
    return c.Command.Run(ctx, f.Args())
  }
  return c.Command.Run(ctx, args)
}
```

Reply:
I can see the `Help()` part is easier. I was hoping we can have a pre-hook to run `f.Parse()` so that every `Run()` implementation doesn't have to explicit call that.